### DISCUSSION TITLE
How to use the output of a command for the suggestions in completion?

### DISCUSSION THREAD

**Discussion by wsw70** [fish-shell/fish-shell#10828] [Q&A] [answered]
I am trying to write a completion but somehow I cannot find my use case neither in the docs nor in previous discussions. Since this is a rather common case I guess I am missing something obvious, sorry for that.

I wrote a cli (`dcs`) and I would like to trigger completion when using the command `dcs start`. The list of the possibilities after `start` are dynamically built with the command `dcs list`.

So my completions should be something like 

    complete -c dcs -a (dcs list)

I then tried to complete `dcs start <tab>`. What I got is below, I use a screenshot because a plain text copy would not show the cursor:

![image](https://github.com/user-attachments/assets/2773113f-ff1b-4f44-a425-313d33d112c7)

- the cursor is on the last but one line, in the column where `start` ends
- the last line is a list of files and directories in the current directory

I tried to further filter the list by typing `dcs start samb<tab>`, the result is even weirder:

![image](https://github.com/user-attachments/assets/5be3c758-918d-419e-b46e-de0b283b9247)

- the command is repeated on the last line
- the end of the last line is inverted (colors)
- the files are gone (probably because nothing matches `samb` there
- the attempt to limit the number of suggestions did not work

**Comment by faho** [comment]
You need to quote the argument to `-a` so it's run when the completion happens, not when the completion script is loaded (and where it will, in this case, most likely error out):

```fish
complete -c dcs -a '(dcs list)'
```

Also this assumes `dcs list` prints the candidates *one per line*. If that's not the case, you need `string split " "`.

**Reply by wsw70** [reply]
> `complete -c dcs -a '(dcs list)'`

I tried that too (to get a quote-delimited string), no change

> Also this assumes `dcs list` prints the candidates one per line

I modified the cli to do that, the problem stays the same, it just makes a column instead of a long line. This is the effect of typing `dcc start <tab>`

![image](https://github.com/user-attachments/assets/80437e0d-55ef-46b1-807e-92f65fdbb1b1)


(`dcc` just generates a shorter list because the one from `dcs` was too long to make a neat screenshot)

For reference, the bare list:

![image](https://github.com/user-attachments/assets/0f60f77d-5e0c-47e1-be85-014dc8a8da4f)

**Reply by faho** [reply, marked as answer]
Okay, run `dcc list >/dev/null`. Does that still print anything?

Did you unload the old completions (restart fish or `complete --erase dcc`)?

**Reply by wsw70** [reply]
Ahhhh.... I know. 

The CLI is in Go, I wrote it right after doing some Python and automatically used `print()` which was accepted (this is some kind of half-assed debug function that should not be there) 😐
`print()` sends it data to stderr.... So yes, when redirecting stdout to `/dev/null` there was still some output.

I switched to `fmt.Write()` which is the proper function an, behold, it works :)

**Thank you very much for the help.**

Now - it works too well so to speak because it also suggests files and directories:

![image](https://github.com/user-attachments/assets/545cb1c9-3c30-462c-b37d-f971aae56c2e)

**Reply by faho** [reply]
> Now - it works too well so to speak because it also suggests files and directories:

`complete -c dcs -f -a '(dcs list)'`

(that's `-f` for `--no-files`, which suppresses the default file completion)

**Reply by wsw70** [reply]
> (that's `-f` for `--no-files`, which suppresses the default file completion)

Fantastic - thanks again!
